File: packages/python/tests/helpers/source_scan.py
"""Source-scanning helpers for the contracts this port's marker is held to.

CC1 (07-cc1-probe-purity): the probe path imports nothing that can do I/O,
transitively, so the probe cannot do I/O even by accident.
CC9 (10-cc9-marker-freeze): the marker attribute name appears as a frozen
literal exactly once in the package, never assembled at run time, never
aliased under a second name.

These live in the test tree on purpose: the scan needs the filesystem, and a
module under `comprehendo/` that imported `pathlib` would itself break CC1.

The scan is an AST walk rather than the TypeScript side's grep-level lexer,
because Python ships a parser: an assembled name (`'__' + 'comprehendo' +
'__'`) is a `BinOp` of constants, not a constant, and constant-folding it is
what tells a computed marker apart from a frozen one. At RUN time the two are
indistinguishable, which is exactly why the gate is a source scan (the
negative kit's computed-marker fixture is the must-fail proof, and it names
both language forms).
"""

from __future__ import annotations

import ast
import sys
from pathlib import Path
from typing import NamedTuple

from comprehendo import MARKER_ATTR

from .kit import PACKAGE_ROOT

__all__ = [
    "IO_MODULES",
    "MARKER_NAME",
    "SRC_ROOT",
    "MarkerSite",
    "SourceFile",
    "external_imports",
    "find_marker_sites",
    "imports_of",
    "marker_sites_in_package",
    "package_files",
    "read_package_sources",
    "stdlib_names",
    "transitive_import_closure",
]

SRC_ROOT = PACKAGE_ROOT / "comprehendo"

#: The name the scan looks for, taken from the package's own single definition
#: site rather than retyped here. The value that constant must EQUAL is
#: asserted against the negative kit's `marker.frozen.python`, so no test in
#: this port hand-writes the marker spelling either.
MARKER_NAME = MARKER_ATTR

#: Standard-library modules that can touch a file, a socket, or a process.
#: A probe that reaches any of these transitively is not side-effect free.
IO_MODULES = frozenset(
    {
        "asyncio",
        "ftplib",
        "http",
        "io",
        "multiprocessing",
        "os",
        "pathlib",
        "shutil",
        "smtplib",
        "socket",
        "sqlite3",
        "ssl",
        "subprocess",
        "tempfile",
        "urllib",
        "webbrowser",
    }
)


class SourceFile(NamedTuple):
    """One module of the package: its path, its text, and its parsed tree."""

    path: str
    text: str
    tree: ast.Module


class MarkerSite(NamedTuple):
    """One place source names the marker, and how it named it."""

    raw: str
    frozen_literal: bool
    computed: bool


def read_package_sources() -> list[SourceFile]:
    """Every `.py` file under `comprehendo/`, recursively, sorted by path."""
    out: list[SourceFile] = []
    for path in sorted(SRC_ROOT.rglob("*.py")):
        text = path.read_text(encoding="utf-8")
        out.append(SourceFile(path.relative_to(SRC_ROOT).as_posix(), text, ast.parse(text)))
    return out


def _docstring_nodes(tree: ast.Module) -> set[int]:
    """Every string constant that is a docstring, by object id.

    Docstrings are prose about the marker, not a definition of it, exactly as
    the TypeScript scan blanks comments before counting.
    """
    found: set[int] = set()
    for node in ast.walk(tree):
        if isinstance(node, (ast.Module, ast.ClassDef, ast.FunctionDef, ast.AsyncFunctionDef)):
            body = getattr(node, "body", [])
            first = body[0] if body else None
            if (
                isinstance(first, ast.Expr)
                and isinstance(first.value, ast.Constant)
                and isinstance(first.value.value, str)
            ):
                found.add(id(first.value))
    return found


def _fold(node: ast.AST) -> tuple[str | None, bool]:
    """Constant-fold a string expression. Returns (value, was_computed)."""
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return node.value, False
    if isinstance(node, ast.BinOp) and isinstance(node.op, ast.Add):
        left, _ = _fold(node.left)
        right, _ = _fold(node.right)
        if left is None or right is None:
            return None, False
        return left + right, True
    if isinstance(node, ast.JoinedStr):
        parts: list[str] = []
        for value in node.values:
            piece, _ = _fold(value)
            if piece is None:
                return None, False
            parts.append(piece)
        return "".join(parts), True
    if isinstance(node, ast.Call) and len(node.args) == 1:
        func = node.func
        argument = node.args[0]
        # `''.join([...])`, the other way a name gets assembled.
        if (
            isinstance(func, ast.Attribute)
            and func.attr == "join"
            and isinstance(func.value, ast.Constant)
            and isinstance(func.value.value, str)
            and isinstance(argument, (ast.List, ast.Tuple))
        ):
            pieces: list[str] = []
            for element in argument.elts:
                piece, _ = _fold(element)
                if piece is None:
                    return None, False
                pieces.append(piece)
            return func.value.value.join(pieces), True
    return None, False


def find_marker_sites(source: str) -> list[MarkerSite]:
    """Every expression in `source` that names the marker, frozen or computed.

    A bare `__comprehendo__` identifier and the string `"__comprehendo__"` are
    both frozen literals (the attribute and its name). Anything that has to be
    evaluated to become the marker is computed, and computed is the violation.
    """
    tree = ast.parse(source)
    docstrings = _docstring_nodes(tree)
    sites: list[MarkerSite] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Name) and node.id == MARKER_NAME:
            sites.append(MarkerSite(MARKER_NAME, True, False))
            continue
        if isinstance(node, ast.Attribute) and node.attr == MARKER_NAME:
            sites.append(MarkerSite(MARKER_NAME, True, False))
            continue
        if isinstance(node, ast.Constant) and isinstance(node.value, str):
            if node.value == MARKER_NAME and id(node) not in docstrings:
                sites.append(MarkerSite(f'"{MARKER_NAME}"', True, False))
            continue
        value, computed = _fold(node)
        if computed and value == MARKER_NAME:
            sites.append(MarkerSite(ast.unparse(node), False, True))
    return sites


def marker_sites_in_package() -> list[tuple[str, MarkerSite]]:
    """Every marker-naming site in the package, tagged with its module."""
    return [
        (source.path, site)
        for source in read_package_sources()
        for site in find_marker_sites(source.text)
    ]


def imports_of(source: SourceFile) -> set[str]:
    """Every module this file imports, top-level package name only."""
    found: set[str] = set()
    for node in ast.walk(source.tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                found.add(alias.name.split(".")[0])
        elif isinstance(node, ast.ImportFrom):
            if node.level > 0:
                found.add("." + (node.module or ""))
            elif node.module is not None:
                found.add(node.module.split(".")[0])
    return found


def transitive_import_closure(entry: str) -> list[SourceFile]:
    """The package modules reachable from `entry` (e.g. `marker.py`)."""
    by_path = {source.path: source for source in read_package_sources()}
    visited: set[str] = set()
    queue = [entry]
    while queue:
        path = queue.pop(0)
        if path in visited or path not in by_path:
            visited.add(path)
            continue
        visited.add(path)
        for name in imports_of(by_path[path]):
            if name.startswith("."):
                target = f"{name[1:]}.py"
                if target in by_path and target not in visited:
                    queue.append(target)
    return [by_path[path] for path in sorted(visited) if path in by_path]


def external_imports(sources: list[SourceFile]) -> set[str]:
    """Every non-relative module name the given sources import."""
    found: set[str] = set()
    for source in sources:
        found |= {name for name in imports_of(source) if not name.startswith(".")}
    return found


def stdlib_names() -> frozenset[str]:
    """The standard-library module names this interpreter ships."""
    return frozenset(sys.stdlib_module_names)


def package_files() -> list[Path]:
    """Every source file of the package, as paths (for the size gate)."""
    return sorted(SRC_ROOT.rglob("*.py"))

File: packages/core/test/helpers/source-scan.ts
/**
 * Source-scanning helpers for the two cross-cutting contracts this package's
 * marker module is held to.
 *
 * CC1 (07-cc1-probe-purity): core imports none of fs, net, http, dns,
 * child_process, so the probe path cannot do I/O even transitively.
 * CC9 (10-cc9-marker-freeze): `Symbol.for('comprehendo')` appears as a frozen
 * literal exactly once in the package, never computed, never aliased.
 *
 * These live in the test tree on purpose: the scan needs `node:fs`, and a
 * module under `src/` that imports it would itself break CC1.
 */
import { readdirSync, readFileSync } from 'node:fs';
import { join, relative, sep } from 'node:path';

/** Absolute path of the package root (`packages/core`). */
export const PACKAGE_ROOT = join(import.meta.dirname, '..', '..');

/** Absolute path of the package's source tree. */
export const SRC_ROOT = join(PACKAGE_ROOT, 'src');

/** Absolute path of the conformance kit shipped by `@comprehendo/spec`. */
export const KIT_ROOT = join(PACKAGE_ROOT, '..', 'spec', 'kit');

export interface SourceFile {
  /** Path relative to `src/`, POSIX-separated, e.g. `marker.ts`. */
  readonly path: string;
  /** The file's full text, comments included. */
  readonly text: string;
  /** The file's text with every comment blanked out, offsets preserved. */
  readonly code: string;
}

/**
 * The transitive import closure reachable from `entryPath` (relative to
 * `src/`, e.g. `marker.ts`), following only relative specifiers
 * (`./foo.js`, `../bar.js`). This is CC1's actual scope, per the doc's
 * Architecture section: "the scan runs over marker.ts (and any module it
 * imports)", not indiscriminately every file the package happens to
 * contain. A file unrelated to the probe path (docs.ts loading a packed
 * corpus, for instance) is explicitly out of scope, the doc says so:
 * "loading a packed corpus is a Docs Engine [13] concern, not the marker
 * probe's".
 */
export function transitiveImportClosure(entryPath: string): SourceFile[] {
  const all = new Map(readCoreSources().map((source) => [source.path, source]));
  const visited = new Set<string>();
  const queue = [entryPath];

  while (queue.length > 0) {
    const path = queue.shift();
    if (path === undefined || visited.has(path)) continue;
    visited.add(path);
    const source = all.get(path);
    if (source === undefined) continue;

    for (const match of source.code.matchAll(/from\s*['"](\.[^'"]+)['"]/g)) {
      const specifier = match[1] ?? '';
      if (!specifier.startsWith('.')) continue;
      // Import specifiers are ESM-style ('./foo.js'); source files are .ts.
      const resolved = specifier.replace(/\.js$/, '.ts').replace(/^\.\//, '');
      if (!visited.has(resolved)) queue.push(resolved);
    }
  }

  return [...visited]
    .map((path) => all.get(path))
    .filter((source): source is SourceFile => source !== undefined)
    .sort((a, b) => a.path.localeCompare(b.path));
}

/** Every `.ts` file under `src/`, recursively, sorted by path. */
export function readCoreSources(): SourceFile[] {
  return readSourcesUnder(SRC_ROOT);
}

/**
 * Every `.ts` file under an arbitrary source root, recursively, sorted by
 * path. `readCoreSources` is this applied to `SRC_ROOT`; CC6 [27]'s
 * package-wide network scan applies it to a sibling package's `src/` too,
 * which core cannot import (one-way dependency) but can still read as files.
 */
export function readSourcesUnder(root: string): SourceFile[] {
  return walk(root)
    .filter((file) => file.endsWith('.ts'))
    .sort()
    .map((file) => {
      const text = readFileSync(file, 'utf8');
      return {
        path: relative(root, file).split(sep).join('/'),
        text,
        code: blankComments(text),
      };
    });
}

function walk(dir: string): string[] {
  const found: string[] = [];
  for (const item of readdirSync(dir, { withFileTypes: true })) {
    const full = join(dir, item.name);
    if (item.isDirectory()) {
      found.push(...walk(full));
    } else if (item.isFile()) {
      found.push(full);
    }
  }
  return found;
}

/**
 * Replace every line and block comment with spaces, preserving length and line
 * breaks so reported offsets and line numbers still line up. String and
 * template literals are left intact, because the marker scan has to read the
 * argument of a `Symbol.for(...)` call.
 *
 * Known limitation of a lexer this small: a regex literal containing a quote
 * character (`/don't/`) would confuse the string state. Core carries no such
 * literal, and the scan is a grep-level lint, not a parser.
 */
export function blankComments(source: string): string {
  const out = source.split('');
  let state: 'code' | 'single' | 'double' | 'template' | 'line' | 'block' = 'code';
  for (let i = 0; i < source.length; i += 1) {
    const char = source[i] ?? '';
    const next = source[i + 1] ?? '';
    if (state === 'code') {
      if (char === '/' && next === '/') {
        state = 'line';
      } else if (char === '/' && next === '*') {
        state = 'block';
      } else if (char === "'") {
        state = 'single';
      } else if (char === '"') {
        state = 'double';
      } else if (char === '`') {
        state = 'template';
      }
      if (state === 'line' || state === 'block') {
        out[i] = ' ';
        out[i + 1] = ' ';
        i += 1;
      }
      continue;
    }
    if (state === 'line') {
      if (char === '\n') {
        state = 'code';
      } else {
        out[i] = ' ';
      }
      continue;
    }
    if (state === 'block') {
      if (char === '*' && next === '/') {
        out[i] = ' ';
        out[i + 1] = ' ';
        i += 1;
        state = 'code';
      } else if (char !== '\n') {
        out[i] = ' ';
      }
      continue;
    }
    // Inside a string or template literal: skip escapes, watch for the closer.
    if (char === '\\') {
      i += 1;
      continue;
    }
    if (
      (state === 'single' && char === "'") ||
      (state === 'double' && char === '"') ||
      (state === 'template' && char === '`')
    ) {
      state = 'code';
    }
  }
  return out.join('');
}

export interface MarkerCall {
  /** The call exactly as written, e.g. `Symbol.for('comprehendo')`. */
  readonly raw: string;
  /** The call's argument text, untrimmed of nothing but surrounding spaces. */
  readonly argument: string;
  /** True only when the argument is the bare string literal `comprehendo`. */
  readonly frozenLiteral: boolean;
  /** True when the argument names the marker but assembles it at run time. */
  readonly computed: boolean;
}

const MARKER_KEY = 'comprehendo';

/**
 * Find every `Symbol.for(...)` call in a piece of JavaScript or TypeScript
 * source. This is CC9's grep-level lint: one definition site, a frozen
 * literal, nothing assembled at run time.
 */
export function findMarkerCalls(source: string): MarkerCall[] {
  const code = blankComments(source);
  const calls: MarkerCall[] = [];
  const opener = /Symbol\s*\.\s*for\s*\(/g;
  let match: RegExpExecArray | null = opener.exec(code);
  while (match !== null) {
    const argStart = match.index + match[0].length;
    const argEnd = findClosingParen(code, argStart);
    if (argEnd === -1) {
      break;
    }
    const argument = code.slice(argStart, argEnd).trim();
    const frozenLiteral = argument === `'${MARKER_KEY}'` || argument === `"${MARKER_KEY}"`;
    calls.push({
      raw: code.slice(match.index, argEnd + 1),
      argument,
      frozenLiteral,
      computed: !frozenLiteral && !isPlainStringLiteral(argument),
    });
    opener.lastIndex = argEnd + 1;
    match = opener.exec(code);
  }
  return calls;
}

function isPlainStringLiteral(argument: string): boolean {
  return /^'[^']*'$/.test(argument) || /^"[^"]*"$/.test(argument);
}

function findClosingParen(code: string, from: number): number {
  let depth = 1;
  for (let i = from; i < code.length; i += 1) {
    const char = code[i];
    if (char === '(') {
      depth += 1;
    } else if (char === ')') {
      depth -= 1;
      if (depth === 0) {
        return i;
      }
    }
  }
  return -1;
}

/**
 * CC1's forbidden modules, exactly the five the contract names. Anything
 * beyond them (tls, https, worker_threads) is a CC6 concern, not this scan's.
 */
export const FORBIDDEN_MODULES = ['fs', 'net', 'http', 'dns', 'child_process'] as const;

/** Every forbidden module a source file imports, statically or dynamically. */
export function findForbiddenImports(source: string): string[] {
  return findModuleImports(source, FORBIDDEN_MODULES);
}

/**
 * Every module named in `modules` that a source file imports, statically or
 * dynamically. `findForbiddenImports` is this applied to CC1's fixed five;
 * CC6 [27]'s package-wide scan applies it to the wider network-module list
 * instead.
 */
export function findModuleImports(source: string, modules: readonly string[]): string[] {
  const code = blankComments(source);
  const found = new Set<string>();
  for (const name of modules) {
    const specifier = `(?:node:)?${name}(?:/[\\w./-]+)?`;
    const patterns = [
      new RegExp(`from\\s*['"]${specifier}['"]`),
      new RegExp(`import\\s*\\(\\s*['"]${specifier}['"]`),
      new RegExp(`require\\s*\\(\\s*['"]${specifier}['"]`),
      new RegExp(`import\\s*['"]${specifier}['"]`),
    ];
    if (patterns.some((pattern) => pattern.test(code))) {
      found.add(name);
    }
  }
  return [...found];
}

/** Read a JSON file from the conformance kit as `unknown`. */
export function readKitJson(...segments: string[]): unknown {
  return JSON.parse(readFileSync(join(KIT_ROOT, ...segments), 'utf8')) as unknown;
}

/** Read a nested property off an `unknown` value without asserting a shape. */
export function at(value: unknown, ...path: string[]): unknown {
  let cursor = value;
  for (const key of path) {
    if (typeof cursor !== 'object' || cursor === null) {
      return undefined;
    }
    cursor = (cursor as Record<string, unknown>)[key];
  }
  return cursor;
}

/** Read a nested property that must be a string, failing loudly when it is not. */
export function stringAt(value: unknown, ...path: string[]): string {
  const found = at(value, ...path);
  if (typeof found !== 'string') {
    throw new TypeError(`expected a string at ${path.join('.')}, got ${typeof found}`);
  }
  return found;
}
